cirq-web/cirq_ts/src/circuit/grid_circuit.ts

```diff
@@ -0,0 +1,101 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import {Group} from 'three';
+import {GridQubit} from './components/grid_qubit';
+import {Symbol3D, SymbolInformation, Coord} from './components/types';
+
+/**
+ * Class that gathers serialized circuit information
+ * from a Python cirq.Circuit and reconstructs it to be
+ * displayed using three.js
+ */
+export class GridCircuit extends Group {
+  readonly initial_num_moments: number;
+  // The keys of this map are serialized Coord arrays [row, col],
+  // representing the row and column where each GridQubit object
+  // is located.
+  private circuit: Map<string, GridQubit>;
+  private padding_factor: number;
+  /**
+   * Class constructor
+   * @param initial_num_moments The number of moments of the circuit. This
+   * determines the length of all the qubit lines in the diagram.
+   * @param qubits A list of GridCoord objects representing the locations of the
+   * qubits in the circuit.
+   * @param padding_factor A number scaling the distance between meshes.
+   */
+  constructor(
+    initial_num_moments: number,
+    qubits: Coord[],
+    padding_factor = 1
+  ) {
+    super();
+    this.padding_factor = padding_factor;
+    this.initial_num_moments = initial_num_moments;
+    this.circuit = new Map();
+
+    for (const coord of qubits) {
+      this.addQubit(coord.row, coord.col, initial_num_moments);
+    }
+  }
+
+  /**
+   * Adds symbols to the circuit map and renders them in the
+   * 3D scene.
+   * @param list A list of SymbolInformation objects that give instructions
+   * on how to render the operations in 3D.
+   */
+  addSymbolsFromList(list: SymbolInformation[]) {
+    for (const symbol of list) {
+      this.addSymbol(symbol);
+    }
+  }
+
+  private addSymbol(symbol_info: SymbolInformation) {
+    //.get gives a reference to an object, so we're good to
+    // just modify
+    const key = [
+      symbol_info.location_info[0].row,
+      symbol_info.location_info[0].col,
+    ].join(',');
+    const symbol = new Symbol3D(symbol_info, this.padding_factor);
+
+    // In production these issues will never come up, since we will always be given
+    // a valid grid circuit as input. For development purposes, however,
+    // these checks will be useful.
+    if (
+      symbol_info.moment < 0 ||
+      symbol_info.moment > this.initial_num_moments
+    ) {
+      throw new Error(
+        `The SymbolInformation object ${symbol_info} has an invalid moment ${symbol_info.moment}`
+      );
+    }
+    const qubit = this.circuit.get(key);
```

Comment:
This is not a circuit, it is just a list of qubits indexed by their coordinates. I strongly suggest renaming to `this.qubit_map`. Also, instead of string indexing, you could use  `private qubit_map: Map<number, Map<number, GridQubit>>;` and then have a little private method that returns a qubit (or undefined) for a given row, col.

Reply:
Would you prefer the nested map to the string ?

Reply:
Yes!!!

Reply:
Done